### PR TITLE
Static Analysis Engine.

### PR DESCRIPTION
**Pull request by chriseth**
This enhances and refactors the existing static analysis engine so that it can be used to retain knowledge about the state across jumps.

### CODE REVIEW

File: libevmasm/KnownState.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+ * derived from the operation and the equivalence class of its arguments. DUPi, SWAPi and some
+ * arithmetic instructions are used to infer equivalences while these classes are determined.
+ */
+class KnownState
+{
+public:
+	using Id = ExpressionClasses::Id;
+	struct StoreOperation
+	{
+		enum Target { Invalid, Memory, Storage };
```
Comment (LianaHus) [inline libevmasm/KnownState.h]: why this is not enum class?

File: libevmasm/KnownState.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+				assertThrow(info.ret <= 1, InvalidDeposit, "");
+				if (info.ret == 1)
+					setStackElement(
+						m_stackHeight + _item.deposit(),
+						m_expressionClasses->find(_item, arguments, _copyItem)
+					);
+			}
+		}
+		for (int p = m_stackHeight; p > m_stackHeight + _item.deposit(); --p)
+			m_stackElements.erase(p);
```
Comment (LianaHus) [inline libevmasm/KnownState.cpp]: you can use erase by range here

File: libevmasm/KnownState.h (hunk 1 above)
Comment (chriseth) [inline libevmasm/KnownState.h]: It only spoils the namespace of the struct, so I would say it is ok.

File: libevmasm/KnownState.cpp (hunk 2 above)
Comment (chriseth) [inline libevmasm/KnownState.cpp]: Oh yes, erase by range is a very good idea!

**Pull request merged**